metadata-ingestion/src/datahub/utilities/logging_manager.py

```diff
@@ -0,0 +1,194 @@
+
+    def _formatMessageColor(self, record: logging.LogRecord) -> str:
+        # Mimic our default format, but with color.
+        message_fg = self.MESSAGE_COLORS.get(record.levelname)
+        return (
+            f'{click.style(f"[{self.formatTime(record, self.datefmt)}]", fg="green", dim=True)} '
+            f"{click.style(f'{record.levelname:8}', fg=message_fg)} "
+            f'{click.style(f"{{{record.name}:{record.lineno}}}", fg="blue", dim=True)} - '
+            f"{click.style(record.getMessage(), fg=message_fg)}"
+        )
+
+
+class _DatahubLogFilter(logging.Filter):
+    def __init__(self, debug: bool) -> None:
+        self.debug = debug
+
+    def filter(self, record: logging.LogRecord) -> bool:
+        top_module = record.name.split(".")[0]
+
+        if top_module in DATAHUB_PACKAGES:
+            if self.debug:
+                return record.levelno >= logging.DEBUG
+            else:
+                return record.levelno >= logging.INFO
+        else:
+            if self.debug:
+                return record.levelno >= logging.WARNING
+            else:
+                return record.levelno >= logging.INFO
+
+
+class _LogBuffer:
+    def __init__(self, maxlen: Optional[int] = None) -> None:
+        self._buffer: Deque[str] = collections.deque(maxlen=maxlen)
+
+    def write(self, line: str) -> None:
+        self._buffer.append(line)
+
+    def clear(self) -> None:
+        self._buffer.clear()
+
+    def format_lines(self) -> str:
+        text = "\n".join(self._buffer)
+        if len(self._buffer) == 0:
+            text = "[no logs]"
+        elif len(self._buffer) == self._buffer.maxlen:
+            text = f"[earlier logs truncated...]\n{text}"
+        return text
+
+
+class _BufferLogHandler(logging.Handler):
+    def __init__(self, storage: _LogBuffer) -> None:
+        super().__init__()
+        self._storage = storage
+
+    def emit(self, record: logging.LogRecord) -> None:
+        self._storage.write(self.format(record))
+
+
+def _remove_all_handlers(logger: logging.Logger) -> None:
+    for handler in logger.handlers[:]:
+        logger.removeHandler(handler)
+        handler.close()
+
+
+_log_buffer = _LogBuffer(maxlen=IN_MEMORY_LOG_BUFFER_SIZE)
+
+
+def get_log_buffer() -> _LogBuffer:
+    return _log_buffer
+
+
+_stream_formatter = _ColorLogFormatter()
+_default_formatter = logging.Formatter(BASE_LOGGING_FORMAT)
+
+
+@contextlib.contextmanager
+def configure_logging(debug: bool, log_file: Optional[str] = None) -> Iterator[None]:
+    _log_buffer.clear()
+
+    with contextlib.ExitStack() as stack:
+        # Create stdout handler.
+        stream_handler = logging.StreamHandler()
+        stream_handler.addFilter(_DatahubLogFilter(debug=debug))
+        stream_handler.setFormatter(_stream_formatter)
+
+        # Create file handler.
+        file_handler: logging.Handler
+        if log_file:
+            file = stack.enter_context(open(log_file, "w"))
+            tee = TeeIO(sys.stdout, file)
```

Comment:
I'm curious - why is this Tee and context piece even required? I'd imagine we should be able to create a global handler that directs to a file without requiring this right?

Reply:
we're redirecting stdout here (e.g. click.echo and print statements)

we want those messages to show up both in the terminal and in the log_file, which is why we need the tee